In [2]:
import os

print(os.listdir("Downloads"))

['1-1.pdf', '1-2.pdf', '127.0.0.1.html', '3601388_1858 (1).pdf', '3601388_1858.pdf', 'A STUDY ON E44N2 TOPOSHEET 2.pdf', 'A Study on toposheet E44N2.docx', 'A Study on toposheet E44N2.pdf', 'adhaar.jpg', 'adhaar.pdf', 'Anaconda3-2025.12-2-Windows-x86_64.exe', 'ApplicationforNewFoodSecurityCard-ApplicationForm.pdf', 'bhavana', 'Bhavana.zip', 'bhavanagit.docx', 'CBM-KCG 4942818398.pdf', 'certificate.txt', 'comet_installer_latest (1).exe', 'comet_installer_latest.exe', 'D-Cyber Receipt.pdf', 'Data Analysis.txt', 'Data processing.txt', 'deloitte certificate.pdf', 'desktop.ini', 'eCertificate 3.pdf', 'eCertificate.pdf', 'eclipse-inst-jre-win64.exe', 'FSC Search.html', 'FSC Search_files', 'Full Stack Development Tasks & Instructions — CodeAlpha M1.pdf', 'G-Cyber Receipt.pdf', 'G.Bhavana(Assignment 3).docx', 'G.Bhavana.docx', 'GANAGANI BHAVANA 88661.pdf', 'Ganagani Bhavana-CV.pdf', 'GANAGANI_BHAVANA_FlowCV_Resume_2026-03-28.pdf', 'GANAGANI_BHAVANA_FlowCV_Resume_2026-03-31 (1).pdf', 'GANAGANI_

In [3]:
import pandas as pd

df = pd.read_csv("Downloads/Titanic-Dataset.csv")

df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [4]:
print("Shape:", df.shape)
df.info()

Shape: (891, 12)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [5]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [6]:
# Fill Age with median

df['Age'].fillna(df['Age'].median(), inplace=True)

# Fill Embarked with mode

df['Embarked'].fillna(df['Embarked'].mode()[0], inplace=True)

# Fill Cabin with Unknown

df['Cabin'].fillna('Unknown', inplace=True)

C:\Users\Bhavana\AppData\Local\Temp\ipykernel_23132\2056637528.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Age'].fillna(df['Age'].median(), inplace=True)
C:\Users\Bhavana\AppData\Local\Temp\ipykernel_23132\2056637528.py:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

F

In [7]:
df.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

In [8]:
print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 0


In [9]:
df.drop_duplicates(inplace=True)

In [10]:
df.drop(
    ['PassengerId','Name','Ticket','Cabin'],
    axis=1,
    inplace=True
)

In [11]:
df.head()

,Survived,Pclass,Sex,Age,SibSp,Parch,Fare,Embarked
0,0,3,male,22.0,1,0,7.2500,S
1,1,1,female,38.0,1,0,71.2833,C
2,1,3,female,26.0,0,0,7.9250,S
3,1,1,female,35.0,1,0,53.1000,S
4,0,3,male,35.0,0,0,8.0500,S


In [12]:
df = pd.get_dummies(
    df,
    columns=['Sex','Embarked'],
    drop_first=True
)

df.head()

,Survived,Pclass,Age,SibSp,Parch,Fare,Sex_male,Embarked_Q,Embarked_S
0,0,3,22.0,1,0,7.2500,True,False,True
1,1,1,38.0,1,0,71.2833,False,False,False
2,1,3,26.0,0,0,7.9250,False,False,True
3,1,1,35.0,1,0,53.1000,False,False,True
4,0,3,35.0,0,0,8.0500,True,False,True


In [13]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

cols = ['Pclass','Age','SibSp','Parch','Fare']

df[cols] = scaler.fit_transform(df[cols])

df.head()

,Survived,Pclass,Age,SibSp,Parch,Fare,Sex_male,Embarked_Q,Embarked_S
0,0,0.827377,-0.565736,0.432793,-0.473674,-0.502445,True,False,True
1,1,-1.566107,0.663861,0.432793,-0.473674,0.786845,False,False,False
2,1,0.827377,-0.258337,-0.474545,-0.473674,-0.488854,False,False,True
3,1,-1.566107,0.433312,0.432793,-0.473674,0.420730,False,False,True
4,0,0.827377,0.433312,-0.474545,-0.473674,-0.486337,True,False,True


In [14]:
X = df.drop('Survived', axis=1)

y = df['Survived']

print(X.shape)
print(y.shape)

(891, 8)
(891,)


In [15]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

print(X_train.shape)
print(X_test.shape)

(712, 8)
(179, 8)


In [16]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

lr = LogisticRegression()

lr.fit(X_train, y_train)

pred_lr = lr.predict(X_test)

print("Logistic Regression Accuracy:",
      accuracy_score(y_test, pred_lr))

Logistic Regression Accuracy: 0.8100558659217877


In [17]:
from sklearn.tree import DecisionTreeClassifier

dt = DecisionTreeClassifier(random_state=42)

dt.fit(X_train, y_train)

pred_dt = dt.predict(X_test)

print("Decision Tree Accuracy:",
      accuracy_score(y_test, pred_dt))

Decision Tree Accuracy: 0.776536312849162


In [18]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(
    n_estimators=100,
    random_state=42
)

rf.fit(X_train, y_train)

pred_rf = rf.predict(X_test)

print("Random Forest Accuracy:",
      accuracy_score(y_test, pred_rf))

Random Forest Accuracy: 0.8156424581005587


In [19]:
print("Logistic Regression :", accuracy_score(y_test, pred_lr))
print("Decision Tree       :", accuracy_score(y_test, pred_dt))
print("Random Forest       :", accuracy_score(y_test, pred_rf))

Logistic Regression : 0.8100558659217877
Decision Tree       : 0.776536312849162
Random Forest       : 0.8156424581005587


In [20]:
from sklearn.metrics import classification_report

print(classification_report(y_test, pred_rf))

              precision    recall  f1-score   support

           0       0.83      0.86      0.85       105
           1       0.79      0.76      0.77        74

    accuracy                           0.82       179
   macro avg       0.81      0.81      0.81       179
weighted avg       0.81      0.82      0.82       179



In [21]:
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, pred_rf)

print(cm)

[[90 15]
 [18 56]]


In [22]:
print("Logistic Regression :", accuracy_score(y_test, pred_lr))
print("Decision Tree       :", accuracy_score(y_test, pred_dt))
print("Random Forest       :", accuracy_score(y_test, pred_rf))

Logistic Regression : 0.8100558659217877
Decision Tree       : 0.776536312849162
Random Forest       : 0.8156424581005587


In [23]:
from sklearn.metrics import classification_report

print(classification_report(y_test, pred_rf))

              precision    recall  f1-score   support

           0       0.83      0.86      0.85       105
           1       0.79      0.76      0.77        74

    accuracy                           0.82       179
   macro avg       0.81      0.81      0.81       179
weighted avg       0.81      0.82      0.82       179



# Conclusion

This project used the Titanic Dataset to predict passenger survival.

Data preprocessing steps included:
- Handling missing values
- Removing duplicate records
- Encoding categorical variables
- Feature scaling
- Train-test splitting

Three machine learning algorithms were applied:

1. Logistic Regression
2. Decision Tree Classifier
3. Random Forest Classifier

The performance of each model was evaluated using accuracy, precision, recall, F1-score, and confusion matrix.

Results showed that the Random Forest Classifier achieved the highest accuracy of 81.56%, making it the best-performing model for this dataset.

Therefore, Random Forest was selected as the final model for Titanic survival prediction.